feat(nuget): add support for updating lock files

fgreinacher opened the pull request:
Resolves #6610

fgreinacher commented:
OK, this is ready for review now @viceice

fgreinacher commented:
(And I think we need to wait for the Docker image, right?)

viceice commented:
Yes, sorry.

fgreinacher commented:
> Yes, sorry.

Not meant as offense, just wasn't sure whether something happened that I have not noticed 🤣

viceice commented:
Yes, too much work on daily job happened. 🙈

JamieMagee commented on lib/manager/nuget/artifacts.ts:
```diff
@@ -0,0 +1,129 @@
+}: UpdateArtifact): Promise<UpdateArtifactsResult[] | null> {
+  logger.debug(`nuget.updateArtifacts(${packageFileName})`);
+
+  if (!/[cs|vb|fs]proj$/i.test(packageFileName)) {
```
```suggestion
  if (!/(?:cs|vb|fs)proj$/i.test(packageFileName)) {
```

Same as the filematch regex

https://github.com/renovatebot/renovate/blob/f52fe033de5fae6be0155c3b36fad1ec6f722f01/lib/manager/nuget/index.ts#L8-L12

JamieMagee commented on lib/manager/nuget/artifacts.ts:
```diff
@@ -0,0 +1,129 @@
+  if (!/[cs|vb|fs]proj$/i.test(packageFileName)) {
+    // This could be implemented in the future if necessary.
+    // It's not that easy though because the questions which
+    // project file to restore how to detmermine which lock files
```
```suggestion
    // project file to restore how to determine which lock files
```

viceice reviewed (commented):
LGTM, but needs dotnet docker images to be published first

viceice commented:
https://hub.docker.com/r/renovate/dotnet/tags

🎉

rarkins commented:
Can someone verify that it works with Docker too?

rarkins reviewed (changes requested):
Looks ok from a code review point of view, should be tested with Docker first before merging

fgreinacher commented:
> Looks ok from a code review point of view, should be tested with Docker first before merging

I'll take care!

fgreinacher commented:
> > Looks ok from a code review point of view, should be tested with Docker first before merging
> 
> I'll take care!

Found a bug (fixed with https://github.com/renovatebot/renovate/pull/7375/commits/68239c6ea99f0433298dda4476e6258567c87c11). 

Otherwise it's working with both [Global](https://github.com/fgreinacher/nuget-lockfile/pull/1/files) and [Docker](https://github.com/fgreinacher/nuget-lockfile-2/pull/1/files) binary sources.

fgreinacher commented:
(BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)

viceice commented:
> (BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)

This will never work on windows because the host and container path need to match. A workaround is to mounbt the local build into the renovate slim image and use that to run renovate in container. Then you are able to mount /tmp to renovate, which will be used as shared volume

fgreinacher commented:
> > (BTW: I first tried this on Windows, where I ran into several problems with the Docker binary source. I ignored this and switched to my Mac, assuming that it's OK like that. Tell me if this should actually work, I'd then dig a bitter deeper to see if I can write a proper issue)
> 
> This will never work on windows because the host and container path need to match. A workaround is to mounbt the local build into the renovate slim image and use that to run renovate in container. Then you are able to mount /tmp to renovate, which will be used as shared volume

Yeah, that's why I just switched to a proper machine 😁. Just wanted to make sure I didn't miss anything essential.

viceice commented:
Alternate would be to use wsl on windows 😉
